### DISCUSSION TITLE
[Question][Hw3] Encoding parameter of PHI node

### DISCUSSION THREAD

**Discussion by retroinspect** [prosyslab-classroom/cs424-program-reasoning#332] [Q&A]
From README.md, I tried to encode following rules using `Z3.Expr.substitute_one` with replacing parameter of current block (`end` in example).
```
R1: then(...) /\ I1 /\ ... /\ In => end(0, ...) 
R2: else(...) /\ J1 /\ ... /\ Jn => end(1, ...)
```

But it results in nonsense counterexample, ignoring original parameter of current block.

What does it mean by above rule that first parameter of `end` was replaced 0 or 1?

**Comment by Allgot** [comment]
I think, since the given example program only contains PHI instruction and return, the rule has been shortened.

So, the full rule for one PHI instruction will be like 'for each prev node, the instruction's value will be the given value'.

For example, let's assume following instruction:
```
%a = phi i32 [1, %node1], [2, %node2], [3, %node3]
```

Then, the rule for this will be
```
R1: node1(...) /\ I1 /\ ... /\ In /\ (%a = 1)
R2: node2(...) /\ J2 /\ ... /\ Jn /\ (%a = 2)
R3: node3(...) /\ K1 /\ ... /\ Kn /\ (%a = 3)
```

As the given example in readme, if this instruction is followed by the return, these become
```
R1: node1(...) /\ I1 /\ ... /\ In /\ (%a = 1) => end(%a)
R2: node2(...) /\ J2 /\ ... /\ Jn /\ (%a = 2) => end(%a)
R3: node3(...) /\ K1 /\ ... /\ Kn /\ (%a = 3) => end(%a)
```

While the verification, if the pred node was node 1, rules may contain 'node1(...)'. It will satisify R1's condition and end(1) will be reachable in turn.

**Comment by Allgot** [comment]
Btw, I don't know why things like I1, ..., In are required. Aren't they implied by the pred node condition?

In other words, the rules' first constraints 'node_(_)' already contains that I1, J1, ..., In, Jn are hold I think...

Are these equivalent?
```
R1: node1(...) /\ I1 /\ ... /\ In /\ (%a = 1)
R1': node1(...) /\ (%a = 1)
```

**Comment by retroinspect** [comment]
Thanks for comment.

I think node1(...) only means that node1 is reachable.
So we need to specify `I1, .... In` to limit the scope of variables used by current block (`end` in the above example)

**Comment by Allgot** [comment]
Oh, actually that makes sense and the following example exhibits it well!

```
node1:
    %a =@ input();
    %cmp = icmp sgt %a 1
    %conv = zext %cmp
    @assume(%conv);
    ...

phi_node:
    %b = phi [%a, %node1], [1, %node2]
    ret %b
```

Without In's fron node1, %a will not be something larger than 1! Thank you :)

**Comment by retroinspect** [comment]
> I think, since the given example program only contains PHI instruction and return, the rule has been shortened.
> 
> So, the full rule for one PHI instruction will be like 'for each prev node, the instruction's value will be the given value'.
> 
> For example, let's assume following instruction:
> 
> ```
> %a = phi i32 [1, %node1], [2, %node2], [3, %node3]
> ```
> 
> Then, the rule for this will be
> 
> ```
> R1: node1(...) /\ I1 /\ ... /\ In /\ (%a = 1)
> R2: node2(...) /\ J2 /\ ... /\ Jn /\ (%a = 2)
> R3: node3(...) /\ K1 /\ ... /\ Kn /\ (%a = 3)
> ```
> 
> As the given example in readme, if this instruction is followed by the return, these become
> 
> ```
> R1: node1(...) /\ I1 /\ ... /\ In /\ (%a = 1) => end(%a)
> R2: node2(...) /\ J2 /\ ... /\ Jn /\ (%a = 2) => end(%a)
> R3: node3(...) /\ K1 /\ ... /\ Kn /\ (%a = 3) => end(%a)
> ```
> 
> While the verification, if the pred node was node 1, rules may contain 'node1(...)'. It will satisify R1's condition and end(1) will be reachable in turn.



> I think, since the given example program only contains PHI instruction and return, the rule has been shortened.
> 
> So, the full rule for one PHI instruction will be like 'for each prev node, the instruction's value will be the given value'.
> 
> For example, let's assume following instruction:
> 
> ```
> %a = phi i32 [1, %node1], [2, %node2], [3, %node3]
> ```
> 
> Then, the rule for this will be
> 
> ```
> R1: node1(...) /\ I1 /\ ... /\ In /\ (%a = 1)
> R2: node2(...) /\ J2 /\ ... /\ Jn /\ (%a = 2)
> R3: node3(...) /\ K1 /\ ... /\ Kn /\ (%a = 3)
> ```
> 
> As the given example in readme, if this instruction is followed by the return, these become
> 
> ```
> R1: node1(...) /\ I1 /\ ... /\ In /\ (%a = 1) => end(%a)
> R2: node2(...) /\ J2 /\ ... /\ Jn /\ (%a = 2) => end(%a)
> R3: node3(...) /\ K1 /\ ... /\ Kn /\ (%a = 3) => end(%a)
> ```
> 
> While the verification, if the pred node was node 1, rules may contain 'node1(...)'. It will satisify R1's condition and end(1) will be reachable in turn.

1.
If you meant %y instead of %a for paramenter (like below formula), it may not hold if any instruction of phi nodes uses %y.
```
R1: node1(...) /\ I1 /\ ... /\ In /\ (%a = 1) => end(%y)
R2: node2(...) /\ J2 /\ ... /\ Jn /\ (%a = 2) => end(%y)
R3: node3(...) /\ K1 /\ ... /\ Kn /\ (%a = 3) => end(%y)
```

For example (like in example3.ll), `%j.0` is determined by `%inc` but for `%inc`'s calculation on `%while.body` needs `%j.0`.
```
while.cond:                                       ; preds = %while.body, %entry
...
  %j.0 = phi i32 [ 0, %entry ], [ %inc, %while.body ]


while.body:                                       ; preds = %while.cond
  %inc = add nsw i32 %j.0, 1
...
```

Therefore, below formula becomes always true since the assumption cannot be true.
```
while.body(...) /\ (%inc = %j.0 + 1) ....  /\  (%j.0 = %inc) => while.cond(%j.0)
```

2.
If you meant introducing new variable for phi parameter, it has another problem.
Z3 fails to relate the original variable of reachability formula and new variable.
For example, here we introduce %j.1 for indicating phi parameter corresponding to %j.0.
And there must be other clauses from `while.cond(%j.0, ...)`.

```
while.body(...) /\ (%inc = %j.0 + 1) ....  /\  (%j.1 = %inc) => while.cond(%j.1)
while.cond(%j.0, ...) /\ (%cmp4 = %j.0 < %call) /\ .... => ....
```
Then Z3 provides counterexample which assigns %j.0 an unrelated value from %j.1, always fails to verifying the given program.

**Comment by Allgot** [comment]
Well, actually I used the rule excluding prev node's clauses(something like R1: prev_node(...) /\ (%phi = value)) and it worked well.

The rationale behind this was if the prev node is reachable, then it must been already constrained by that node.

So, from this example,
```
node1: 
    %a =@ input(); 
    %cmp = icmp sgt %a 1 
    %conv = zext %cmp 
    @assume(%conv); 
    ... 

phi_node: 
    %b = phi [%a, %node1], [1, %node2] 
    ret %b
```

I just made a rule R1as following:
```
R1: node1(...) /\ (%b = %a) => end(%b)
```

I think, without node1's clauses, %a should be greater than 1 if the node1 is reachable.. isn't it??

**Comment by Allgot** [comment]
> > I think, since the given example program only contains PHI instruction and return, the rule has been shortened.
> > 
> > So, the full rule for one PHI instruction will be like 'for each prev node, the instruction's value will be the given value'.
> > 
> > For example, let's assume following instruction:
> > 
> > ```
> > %a = phi i32 [1, %node1], [2, %node2], [3, %node3]
> > ```
> > 
> > Then, the rule for this will be
> > 
> > ```
> > R1: node1(...) /\ I1 /\ ... /\ In /\ (%a = 1)
> > R2: node2(...) /\ J2 /\ ... /\ Jn /\ (%a = 2)
> > R3: node3(...) /\ K1 /\ ... /\ Kn /\ (%a = 3)
> > ```
> > 
> > As the given example in readme, if this instruction is followed by the return, these become
> > 
> > ```
> > R1: node1(...) /\ I1 /\ ... /\ In /\ (%a = 1) => end(%a)
> > R2: node2(...) /\ J2 /\ ... /\ Jn /\ (%a = 2) => end(%a)
> > R3: node3(...) /\ K1 /\ ... /\ Kn /\ (%a = 3) => end(%a)
> > ```
> > 
> > While the verification, if the pred node was node 1, rules may contain 'node1(...)'. It will satisify R1's condition and end(1) will be reachable in turn.
> 
> 
> 
> > I think, since the given example program only contains PHI instruction and return, the rule has been shortened.
> > 
> > So, the full rule for one PHI instruction will be like 'for each prev node, the instruction's value will be the given value'.
> > 
> > For example, let's assume following instruction:
> > 
> > ```
> > %a = phi i32 [1, %node1], [2, %node2], [3, %node3]
> > ```
> > 
> > Then, the rule for this will be
> > 
> > ```
> > R1: node1(...) /\ I1 /\ ... /\ In /\ (%a = 1)
> > R2: node2(...) /\ J2 /\ ... /\ Jn /\ (%a = 2)
> > R3: node3(...) /\ K1 /\ ... /\ Kn /\ (%a = 3)
> > ```
> > 
> > As the given example in readme, if this instruction is followed by the return, these become
> > 
> > ```
> > R1: node1(...) /\ I1 /\ ... /\ In /\ (%a = 1) => end(%a)
> > R2: node2(...) /\ J2 /\ ... /\ Jn /\ (%a = 2) => end(%a)
> > R3: node3(...) /\ K1 /\ ... /\ Kn /\ (%a = 3) => end(%a)
> > ```
> > 
> > While the verification, if the pred node was node 1, rules may contain 'node1(...)'. It will satisify R1's condition and end(1) will be reachable in turn.
> 
> 1.
> If you meant %y instead of %a for paramenter (like below formula), it may not hold if any instruction of phi nodes uses %y.
> ```
> R1: node1(...) /\ I1 /\ ... /\ In /\ (%a = 1) => end(%y)
> R2: node2(...) /\ J2 /\ ... /\ Jn /\ (%a = 2) => end(%y)
> R3: node3(...) /\ K1 /\ ... /\ Kn /\ (%a = 3) => end(%y)
> ```
> 
> For example (like in example3.ll), `%j.0` is determined by `%inc` but for `%inc`'s calculation on `%while.body` needs `%j.0`.
> ```
> while.cond:                                       ; preds = %while.body, %entry
> ...
>   %j.0 = phi i32 [ 0, %entry ], [ %inc, %while.body ]
> 
> 
> while.body:                                       ; preds = %while.cond
>   %inc = add nsw i32 %j.0, 1
> ...
> ```
> 
> Therefore, below formula becomes always true since the assumption cannot be true.
> ```
> while.body(...) /\ (%inc = %j.0 + 1) ....  /\  (%j.0 = %inc) => while.cond(%j.0)
> ```
> 
> 2.
> If you meant introducing new variable for phi parameter, it has another problem.
> Z3 fails to relate the original variable of reachability formula and new variable.
> For example, here we introduce %j.1 for indicating phi parameter corresponding to %j.0.
> And there must be other clauses from `while.cond(%j.0, ...)`.
> 
> ```
> while.body(...) /\ (%inc = %j.0 + 1) ....  /\  (%j.1 = %inc) => while.cond(%j.1)
> while.cond(%j.0, ...) /\ (%cmp4 = %j.0 < %call) /\ .... => ....
> ```
> Then Z3 provides counterexample which assigns %j.0 an unrelated value from %j.1, always fails to verifying the given program.
> 
> 

From your example, in while.body also uses %j.0 which is not constrained in that block. Rather, it is already constrained by its prev node, while.cond.

Thus, I think only specifying the reachability condition is enough for phi instruction.
